### PR TITLE
Update snapshot controller

### PR DESCRIPTION
**Pull request by xing-yang**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR updates the snapshot controller logic.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/common-controller/snapshot_controller.go [hunk 1]
```diff
@@ -1193,3 +1218,48 @@ func (ctrl *csiSnapshotCommonController) contentExists(snapshot *crdv1.VolumeSna
 	// Found in content cache store and convert object is successful
 	return content, nil
 }
+
+func (ctrl *csiSnapshotCommonController) snapshotExists(snapshotName string) (*crdv1.VolumeSnapshot, error) {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Change it to getSnapshotFromStore.

File: pkg/common-controller/snapshot_controller.go [hunk 2]
```diff
@@ -206,18 +191,29 @@ func (ctrl *csiSnapshotCommonController) processFinalizersAndCheckandDeleteConte
...
-	err = ctrl.checkandRemoveSnapshotFinalizersAndCheckandDeleteContent(snapshot, content, deleteContent)
+// processIfDeletionTimestampSet processes finalizers and deletes the content when appropriate
+// It checks if contents exists, it checks if snapshot has bi-directional binding, it checks if
+// finalizers should be removed, and it checks if content should be deleted and deletes it if needed.
+func (ctrl *csiSnapshotCommonController) processIfDeletionTimestampSet(snapshot *crdv1.VolumeSnapshot) error {
+	klog.V(5).Infof("processIfDeletionTimestampSet VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
+
+	content, deleteContent, _, err := ctrl.checkContentAndBoundStatus(snapshot)
 	if err != nil {
 		return err
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: "no content is found" will not return err.  I need to add better comments here.

```
func (ctrl *csiSnapshotCommonController) getContentFromStore(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshotContent, error) {
        ......
        obj, found, err := ctrl.contentStore.GetByKey(contentName)
        if err != nil {
                return nil, err
        }
        // Not in the content cache store, no error
        if !found {
                return nil, nil
        }
```

File: pkg/common-controller/snapshot_controller.go [hunk 3]
```diff
@@ -141,22 +136,9 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
-				return newControllerUpdateError(content.Name, err.Error())
-			}
-
-			_, err = ctrl.storeContentUpdate(updateContent)
-			if err != nil {
-				klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", content.Name, err)
-				return err
-			}
-			klog.V(5).Infof("syncContent: volume snapshot content %+v", content)
+		_, err = ctrl.setAnnVolumeSnapshotBeingDeleted(content)
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Fixed

File: pkg/common-controller/snapshot_controller.go [hunk 4]
```diff
@@ -85,43 +85,38 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
 	if snapshot != nil && snapshot.UID != content.Spec.VolumeSnapshotRef.UID {
 		// The snapshot that the content was pointing to was deleted, and another
 		// with the same name created.
 		klog.V(4).Infof("synchronizing VolumeSnapshotContent[%s]: content %s has different UID, the old one must have been deleted", content.Name, snapshotName)
 		// Treat the content as bound to a missing snapshot.
 		snapshot = nil
 	} else {
+		// TODO(xyang): Write a function to check if snapshot.Status is different from content.Status and add snapshot to queue if there is a difference and it is worth triggering an snapshot status update
 		// Check if content status is set to true and update snapshot status if so
 		if snapshot != nil && content.Status != nil && content.Status.ReadyToUse != nil && *content.Status.ReadyToUse == true {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Ok, will fix soon.

[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Fixed.

File: pkg/common-controller/snapshot_controller.go [hunk 5]
```diff
@@ -281,10 +283,11 @@ func (ctrl *csiSnapshotCommonController) checkandAddSnapshotFinalizers(snapshot
...
 		ctrl.addSnapshotFinalizer(snapshot, addSourceFinalizer, addBoundFinalizer)
 	}
+	return nil
 }
 
 // syncReadySnapshot checks the snapshot which has been bound to snapshot content successfully before.
-// If there is any problem with the binding (e.g., snapshot points to a non-exist snapshot content), update the snapshot status and emit event.
+// If there is any problem with the binding (e.g., snapshot points to a non-existent snapshot content), update the snapshot status and emit event.
 func (ctrl *csiSnapshotCommonController) syncReadySnapshot(snapshot *crdv1.VolumeSnapshot) error {
 	if !utils.IsBoundVolumeSnapshotContentNameSet(snapshot) {
```
Comment (yuxiangqian) [inline pkg/common-controller/snapshot_controller.go]: is this check necessary? or if it's needed, it rather should be an error instead of returning nil directly?

File: pkg/common-controller/snapshot_controller.go [hunk 6]
```diff
@@ -171,31 +153,34 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
+			return err
+		}
+	} else {
+		klog.V(5).Infof("syncSnapshot[%s]: check if we should add finalizers on snapshot", utils.SnapshotKey(snapshot))
+		ctrl.checkandAddSnapshotFinalizers(snapshot)
+		// Need to build or update snapshot.Status in following cases:
+		// 1) snapshot.Status is nil
+		// 2) snapshot.Status.ReadyToUse is false
+		// 3) snapshot.Status.BoundVolumeSnapshotContentName is not set
+		if !utils.IsSnapshotReady(snapshot) || !utils.IsBoundVolumeSnapshotContentNameSet(snapshot) {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Yes, we should update status and continue to sync up if BoundVolumeSnapshotContentNameSet is not set.

File: pkg/common-controller/snapshot_controller.go [hunk 7]
```diff
@@ -171,31 +153,34 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
+			return ctrl.syncUnreadySnapshot(snapshot)
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: That will be divided into several cases below:
1) If "snapshot.Status != nil && snapshot.Status.BoundVolumeSnapshotContentName != nil && snapshot.Status.Ready == nil", call syncUnreadySnapshot().
2) If "snapshot.Status != nil && snapshot.Status.BoundVolumeSnapshotContentName != nil && snapshot.Status.Ready != nil && *snapshot.Status.Ready == false", call syncUnreadySnapshot().
3) If "snapshot.Status != nil && snapshot.Status.BoundVolumeSnapshotContentName != nil && snapshot.Status.Ready != nil && *snapshot.Status.Ready == true", call syncReadySnapshot().

File: pkg/common-controller/snapshot_controller.go [hunk 8]
```diff
@@ -206,18 +191,30 @@ func (ctrl *csiSnapshotCommonController) processFinalizersAndCheckandDeleteConte
...
 	}
+	return content, deleteContent, snapshotBound, nil
+}
 
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent[%s]: delete snapshot content and remove finalizer from snapshot if needed", utils.SnapshotKey(snapshot))
-	err = ctrl.checkandRemoveSnapshotFinalizersAndCheckandDeleteContent(snapshot, content, deleteContent)
+// processIfDeletionTimestampSet processes finalizers and deletes the content when appropriate
+// It checks if contents exists, it checks if snapshot has bi-directional binding, it checks if
+// finalizers should be removed, and it checks if content should be deleted and deletes it if needed.
+func (ctrl *csiSnapshotCommonController) processIfDeletionTimestampSet(snapshot *crdv1.VolumeSnapshot) error {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: If snapshot content has not been created yet, user can't create a PVC from the snapshot.  The provisioner will check whether snapshot and content has bi-directional binding and snapshot status readyToUse is true before proceeding to create PVC from snapshot.

File: pkg/common-controller/snapshot_controller.go [hunk 9]
```diff
@@ -241,7 +238,7 @@ func (ctrl *csiSnapshotCommonController) checkandRemoveSnapshotFinalizersAndChec
 		// Check if a volume is being created from snapshot.
 		inUse := ctrl.isVolumeBeingCreatedFromSnapshot(snapshot)
 
-		klog.V(5).Infof("syncSnapshot[%s]: set DeletionTimeStamp on content.", utils.SnapshotKey(snapshot))
+		klog.V(5).Infof("checkandRemoveSnapshotFinalizersAndCheckandDeleteContent[%s]: set DeletionTimeStamp on content.", utils.SnapshotKey(snapshot))
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: "IsSnapshotDeletionCandidate" only checks if deletionTimeStamp is set on VolumeSnapshot and if there are finalizers on VolumeSnapshot.  
checkandRemoveSnapshotFinalizersAndCheckandDeleteContent calls IsSnapshotDeletionCandidate, and then checks whether those finalizers can be removed by checking if a snapshot is bound to a content and if a volume is being created from the snapshot.  If snapshot is not bound and not used by a volume, finalizers will be removed.

File: pkg/common-controller/snapshot_controller.go (hunk 5 above)
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Yeah, this should be an error.  Will fix it.

File: pkg/utils/util.go [hunk 10]
```diff
@@ -319,23 +319,23 @@ func NoResyncPeriodFunc() time.Duration {
...
-// needToAddSnapshotAsSourceFinalizer checks if a Finalizer needs to be added for the volume snapshot as a source for PVC.
+// NeedToAddSnapshotAsSourceFinalizer checks if a Finalizer needs to be added for the volume snapshot as a source for PVC.
 func NeedToAddSnapshotAsSourceFinalizer(snapshot *crdv1.VolumeSnapshot) bool {
 	return snapshot.ObjectMeta.DeletionTimestamp == nil && !slice.ContainsString(snapshot.ObjectMeta.Finalizers, VolumeSnapshotAsSourceFinalizer, nil)
 }
 
-// needToAddSnapshotBoundFinalizer checks if a Finalizer needs to be added for the bound volume snapshot.
+// NeedToAddSnapshotBoundFinalizer checks if a Finalizer needs to be added for the bound volume snapshot.
 func NeedToAddSnapshotBoundFinalizer(snapshot *crdv1.VolumeSnapshot) bool {
 	return snapshot.ObjectMeta.DeletionTimestamp == nil && !slice.ContainsString(snapshot.ObjectMeta.Finalizers, VolumeSnapshotBoundFinalizer, nil) && snapshot.Status != nil && snapshot.Status.BoundVolumeSnapshotContentName != nil
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/utils/util.go]: Good point.  Will change.

File: pkg/common-controller/snapshot_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: discussed around this offline, we might be able to merge this case with the first case, I will leave it to @xing-yang to think further :)

[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: I looked at getMatchSnapshotContent(), it is actually not searching in the API server.  It is searching in ctrl.contentStore.List() which is the cache store.

I agree that ctrl.contentStore.GetByKey() is still more efficient, so I made the fix.

File: pkg/common-controller/snapshot_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Yes, usually it should not happen.  This is to ensure that for whatever reason this BoundVolumeSnapshotContentName field got removed, we still try to update snapshot status and bring it back.

File: pkg/common-controller/snapshot_controller.go [hunk 11]
```diff
@@ -85,78 +85,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
-
-			_, err = ctrl.storeContentUpdate(updateContent)
-			if err != nil {
-				klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", content.Name, err)
-				return err
-			}
-			klog.V(5).Infof("syncContent: volume snapshot content %+v", content)
+	// due to the finalizer.
+	if snapshot != nil && utils.IsSnapshotDeletionCandidate(snapshot) {
+		err = ctrl.setAnnVolumeSnapshotBeingDeleted(content)
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Done

File: pkg/common-controller/snapshot_controller.go [hunk 12]
```diff
@@ -892,6 +893,39 @@ func (ctrl *csiSnapshotCommonController) bindandUpdateVolumeSnapshot(snapshotCon
...
 
+// needsUpdateSnapshotStatus compares snapshot status with the content status and decide
+// if snapshot status needs to be updated based on content status
+func (ctrl *csiSnapshotCommonController) needsUpdateSnapshotStatus(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) bool {
+	klog.V(5).Infof("needsUpdateSnapshotStatus[%s]", utils.SnapshotKey(snapshot))
+
+	if snapshot.Status == nil && content.Status != nil {
+		return true
+	}
+	if snapshot.Status != nil && content.Status == nil {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Done

File: pkg/common-controller/snapshot_controller.go [hunk 13]
```diff
@@ -892,6 +893,39 @@ func (ctrl *csiSnapshotCommonController) bindandUpdateVolumeSnapshot(snapshotCon
...
+	if snapshot.Status.CreationTime == nil && content.Status.CreationTime != nil {
+		return true
+	}
+	if snapshot.Status.ReadyToUse == nil && content.Status.ReadyToUse != nil {
+		return true
+	}
+	if snapshot.Status.ReadyToUse != nil && content.Status.ReadyToUse != nil && snapshot.Status.ReadyToUse != content.Status.ReadyToUse {
+		return true
+	}
+	if (snapshot.Status.RestoreSize == nil && content.Status.RestoreSize != nil) || (snapshot.Status.RestoreSize != nil && snapshot.Status.RestoreSize.IsZero() && content.Status.RestoreSize != nil && *content.Status.RestoreSize > 0) {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Done

File: pkg/common-controller/snapshot_controller.go [hunk 14]
```diff
@@ -171,31 +161,34 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
-// finalizers should be added or removed, and it checks if content should be deleted and deletes it
-// if needed.
-func (ctrl *csiSnapshotCommonController) processFinalizersAndCheckandDeleteContent(snapshot *crdv1.VolumeSnapshot) error {
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
-
+func (ctrl *csiSnapshotCommonController) checkContentAndBoundStatus(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshotContent, bool, bool, error) {
 	// If content is deleted already, remove SnapshotBound finalizer
-	content, err := ctrl.contentExists(snapshot)
+	content, err := ctrl.getContentFromStore(snapshot)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Done

File: pkg/common-controller/snapshot_controller.go [hunk 15]
```diff
@@ -241,7 +246,7 @@ func (ctrl *csiSnapshotCommonController) checkandRemoveSnapshotFinalizersAndChec
 		// Check if a volume is being created from snapshot.
 		inUse := ctrl.isVolumeBeingCreatedFromSnapshot(snapshot)
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Done

**Comment by xing-yang** [comment]
@yuxiangqian addressed your comments.  PTAL.

File: pkg/common-controller/snapshot_controller.go [hunk 16]
```diff
@@ -85,78 +85,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
 	if content.Spec.VolumeSnapshotRef.UID == "" {
 		klog.V(4).Infof("synchronizing VolumeSnapshotContent[%s]: VolumeSnapshotContent is pre-bound to VolumeSnapshot %s", content.Name, snapshotName)
 		return nil
 	}
-	// Get the VolumeSnapshot by _name_
+
+	if utils.NeedToAddContentFinalizer(content) {
+		// Content is not being deleted -> it should have the finalizer.
+		klog.V(5).Infof("syncContent: Add Finalizer for VolumeSnapshotContent[%s]", content.Name)
+		return ctrl.addContentFinalizer(content)
```
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: I think adding finalizer should trigger content update event, and thenthe content will be re-queue'd to be processed again, that would be when status of snapshot gets updated?

File: pkg/common-controller/snapshot_controller.go [hunk 17]
```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
 	}
+	return content, deleteContent, snapshotBound, nil
+}
 
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent[%s]: delete snapshot content and remove finalizer from snapshot if needed", utils.SnapshotKey(snapshot))
-	err = ctrl.checkandRemoveSnapshotFinalizersAndCheckandDeleteContent(snapshot, content, deleteContent)
+// processIfDeletionTimestampSet processes finalizers and deletes the content when appropriate
+// It checks if contents exists, it checks if snapshot has bi-directional binding, it checks if
+// finalizers should be removed, and it checks if content should be deleted and deletes it if needed.
+func (ctrl *csiSnapshotCommonController) processIfDeletionTimestampSet(snapshot *crdv1.VolumeSnapshot) error {
```
Comment (jingxu97) [inline pkg/common-controller/snapshot_controller.go]: I think the logic implemented here is confusing. There are several layers of calls. Is the following logic sufficient?

if snapshot.DeletionTimeStamp is set (it indicates it must has finalizer)
1. if snapshotAsSource finalizer is set
 --  (double check pvc is still using it) -- do nothing
 -- if pvc finishes restoring, remove finalizer
2. if SnapshotBound finalizer is set
  -- content exist, set deletiontimestamp on content (can also set annotation at the same time here)
  -- context does not exist, delete finalizer

File: pkg/common-controller/snapshot_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/common-controller/snapshot_controller.go]: why not just let it continue to update status after adding finalizer?

File: pkg/common-controller/snapshot_controller.go [hunk 18]
```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
-	if !utils.IsSnapshotReady(snapshot) {
-		return ctrl.syncUnreadySnapshot(snapshot)
+	if snapshot.ObjectMeta.DeletionTimestamp != nil {
+		err := ctrl.processIfDeletionTimestampSet(snapshot)
+		if err != nil {
+			return err
+		}
+	} else {
+		klog.V(5).Infof("syncSnapshot[%s]: check if we should add finalizers on snapshot", utils.SnapshotKey(snapshot))
+		ctrl.checkandAddSnapshotFinalizers(snapshot)
```
Comment (jingxu97) [inline pkg/common-controller/snapshot_controller.go]: each time snapshot sync, it will trigger checkandAddSnapshotFinalizers, which will check and update finalizer even if finalizer is already added.

File: pkg/common-controller/snapshot_controller.go [hunk 19]
```diff
@@ -266,7 +270,12 @@ func (ctrl *csiSnapshotCommonController) checkandRemoveSnapshotFinalizersAndChec
 }
 
 // checkandAddSnapshotFinalizers checks and adds snapshot finailzers when needed
-func (ctrl *csiSnapshotCommonController) checkandAddSnapshotFinalizers(snapshot *crdv1.VolumeSnapshot, snapshotBound bool, deleteContent bool) {
+func (ctrl *csiSnapshotCommonController) checkandAddSnapshotFinalizers(snapshot *crdv1.VolumeSnapshot) error {
```
Comment (jingxu97) [inline pkg/common-controller/snapshot_controller.go]: Looks like the logic is to addSnapshotAsSourceFinalizer no matter whether there is PVC is currently using snapshot. I think it should be fine as long as we have correct logic to remove it when needed. But worth to put a note about it.

File: pkg/common-controller/snapshot_controller.go [hunk 20]
```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
 func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
```
Comment (jingxu97) [inline pkg/common-controller/snapshot_controller.go]: I feel the logic inside this function is kind of hard to follow. Trying to see whether the following top level functions (we can work out details of functions later) can help.

```
func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
	klog.V(5).Infof("synchronizing VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))

	if snapshot.ObjectMeta.DeletionTimestamp != nil {
		return ctrl.syncDeletingSnapshot(snapshot)

	} else {
		klog.V(5).Infof("syncSnapshot[%s]: check if we should add finalizers on snapshot", utils.SnapshotKey(snapshot))
		ctrl.checkandAddSnapshotFinalizers(snapshot)

		if snapshot.Status == nil {
			return ctrl.syncUnBoundSnapshot(snapshot)
		} else {
			return ctrl.syncBoundSnapshot(snapshot)
		}
	}
	return nil
}

func (ctrl *csiSnapshotCommonController) syncUnBoundSnapshot(snapshot *crdv1.VolumeSnapshot) error {
	content, err := ctrl.getContentFromStore(snapshot)
	if err != nil {
		return err
	}
	
	if content != nil {
		addBoundFinalizer(snapshot)
		addSnapshotAsSourceFinalizer(snapshot)
		return updateSnapshotFromContent(snapshot, content)
	} else {
		if isStaticBinding(snapshot) {
			return nil
		}
		addBoundFinalizer(snapshot)
		return createSnapshotContent(snapshot)
	}

}

func (ctrl *csiSnapshotCommonController) syncBoundSnapshot(snapshot *crdv1.VolumeSnapshot) error {
	content, err := ctrl.getContentFromStore(snapshot)
	if err != nil {
		return err
	}
	
	if content == nil {
		return updateSnapshotWithNoContentError(snapshot)
	} else {
		return updateSnapshotFromContent(snapshot, content)
	}
}

```

File: pkg/common-controller/snapshot_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: certainly could update the status of volume snapshot, just syncContent does not really update snapshot status, samething at line 121, it just put an item into the work queue.  I actually prefer to follow the same pattern here and let syncSnapshot handle it.
One thing was not clear to me though, is if there is no VolumeSnapshotContentFinalizer on the content, i.e., NeedToAddContentFinalizer returns false, AND there is no snapshot object, should the finalizer be added?
maybe move this if block after line 121?

File: pkg/common-controller/snapshot_controller.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: Maybe I am wrong, but looks like it will be a no-op if finalizers presents? @xing-yang could you check?

[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/common-controller/snapshot_controller.go]: yes, I thought I deleted this comment already. There is no issue here.

File: pkg/common-controller/snapshot_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: yeah I guess that was your thought as well. Just it seems to be very unlikely, if not impossible, that only the BoundVolumeSnapshotContentName field is lost, if lost, rather the whole status is lost?

File: pkg/common-controller/snapshot_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: there reason why is, if the finalizer is added to the content object regardless, it will make deletion of an orphaned content not easy?

[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/common-controller/snapshot_controller.go]: right, I meant to add snapshot to the queue to trigger update.  With current logic, it takes a little bit longer to wait for controller to process content from the queue again and add snapshot into the queue. Both should work, so I don't have strong objection about it.

1. sync content, add content finalizer, 
2. content added to the queue back, 
3. sync content again, it goes to check snapshot exist logic and put snapshot into the queue.

File: pkg/common-controller/snapshot_controller.go [hunk 21]
```diff
@@ -85,79 +85,66 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
-
-			_, err = ctrl.storeContentUpdate(updateContent)
-			if err != nil {
-				klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", content.Name, err)
-				return err
-			}
-			klog.V(5).Infof("syncContent: volume snapshot content %+v", content)
-		}
+	// due to the finalizer.
+	if snapshot != nil && utils.IsSnapshotDeletionCandidate(snapshot) {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: It checks if deletionTimestamp is set and finalizer is also set.  If true, it sets the annotation.  Then the sidecar controller will check if it is safe to delete the content finalizer and delete the content, etc.
This follows the pattern of the finalizers in the in-tree pv/pvc protection controller.  It makes sure that it tries to add finalizer first to protection the resource before trying to delete it.

File: pkg/common-controller/snapshot_controller.go [hunk 22]
```diff
@@ -85,78 +85,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Removed.

File: pkg/common-controller/snapshot_controller.go [hunk 23]
```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
 }
 
-// processFinalizersAndCheckandDeleteContent processes finalizers and deletes the content when appropriate
-// It checks if contents exists, it checks if snapshot has bi-directional binding, it checks if
-// finalizers should be added or removed, and it checks if content should be deleted and deletes it
-// if needed.
-func (ctrl *csiSnapshotCommonController) processFinalizersAndCheckandDeleteContent(snapshot *crdv1.VolumeSnapshot) error {
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
-
+func (ctrl *csiSnapshotCommonController) checkContentAndBoundStatus(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshotContent, bool, bool, error) {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Added

File: pkg/common-controller/snapshot_controller.go [hunk 24]
```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
...
 		klog.Infof("syncSnapshot: VolumeSnapshot %s is bound to volumeSnapshotContent [%s]", snapshot.Name, content.Name)
 		snapshotBound = true
+
 	}
+	return content, deleteContent, snapshotBound, nil
+}
 
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent[%s]: delete snapshot content and remove finalizer from snapshot if needed", utils.SnapshotKey(snapshot))
-	err = ctrl.checkandRemoveSnapshotFinalizersAndCheckandDeleteContent(snapshot, content, deleteContent)
+// processIfDeletionTimestampSet processes finalizers and deletes the content when appropriate
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Renamed to processSnapshotWithDeletionTimestamp.  "syncDeletingSnapshot" would be confused with other functions such as syncSnapshot.

File: pkg/common-controller/snapshot_controller.go (hunk 19 above)
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Added note before NeedToAddSnapshotAsSourceFinalizer().

File: pkg/common-controller/snapshot_controller.go [hunk 25]
```diff
@@ -892,6 +893,39 @@ func (ctrl *csiSnapshotCommonController) bindandUpdateVolumeSnapshot(snapshotCon
 	return snapshotCopy, nil
 }
 
+// needsUpdateSnapshotStatus compares snapshot status with the content status and decide
+// if snapshot status needs to be updated based on content status
+func (ctrl *csiSnapshotCommonController) needsUpdateSnapshotStatus(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) bool {
```
[the remark this answers is not in the document]
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: Here we are trying to avoid triggering syncSnapshot if not necessary.

File: pkg/common-controller/snapshot_controller.go (hunk 17 above)
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: There are 3 steps in processIfDeletionTimestampSet (renamed to processSnapshotWithDeletionTimestamp):
1. Call a helper function checkContentAndBoundStatus() to check content and bound status.
2. Call checkandRemoveSnapshotFinalizersAndCheckandDeleteContent() with information obtained from step 1.  This function name is very long but the name suggests what it does.  It determines whether to remove finalizers on snapshot and whether to delete content.
3. Call checkandRemovePVCFinalizer() to determine whether to remove the finalizer on PVC.

I added comments to explain that.

File: pkg/common-controller/snapshot_controller.go (hunk 20 above)
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: I initially thought about splitting syncSnapshot by whether snapshot.Status is nil or not, but I found that it will make the logic more complicated to follow.  If snapshot.Status != nil but it is empty or it is not empty but only with boundVolumeSnapshotContentName set but readyToUse is nil or false, and assume content != nil, we'll go ahead to update snapshot status from content status.  However content status could also be nil.  So we'll have to return without any update on snapshot status and wait for the next sync.  Eventually we'll get there but it will take longer.  It will also have duplicate logic to update status spreading in both syncSnapshotWithStatus and syncSnapshotWithoutStatus.

So I think the current logic is more straight-forward.

File: pkg/common-controller/snapshot_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: PTAL, I still think the check is not quite necessary.

[the remark this answers is not in the document]
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: PTAL, I still think the check is not quite necessary.

File: pkg/common-controller/snapshot_controller.go [hunk 26]
```diff
@@ -266,9 +273,20 @@ func (ctrl *csiSnapshotCommonController) checkandRemoveSnapshotFinalizersAndChec
...
+func (ctrl *csiSnapshotCommonController) checkandAddSnapshotFinalizers(snapshot *crdv1.VolumeSnapshot) error {
+	_, deleteContent, snapshotBound, err := ctrl.checkContentAndBoundStatus(snapshot)
+	if err != nil {
+		return err
+	}
+
 	addSourceFinalizer := false
 	addBoundFinalizer := false
+	// NOTE: Source finalizer will be added to snapshot if
+	// DeletionTimeStamp is nil and it is not set yet.
```
Comment (yuxiangqian) [inline pkg/common-controller/snapshot_controller.go]: nit, "it is not set yet" means what has not been set yet, the deletion time stamp?

File: pkg/common-controller/snapshot_controller.go [hunk 27]
```diff
@@ -987,11 +1029,34 @@ func (ctrl *csiSnapshotCommonController) getVolumeFromVolumeSnapshot(snapshot *c
...
+
 	klog.V(5).Infof("getVolumeFromVolumeSnapshot: snapshot [%s] PV name [%s]", snapshot.Name, pvName)
 
 	return pv, nil
 }
 
+// isVolumeBoundToClaim returns true, if given volume is pre-bound or bound
+// to specific claim. Both claim.Name and claim.Namespace must be equal.
+// If claim.UID is present in volume.Spec.ClaimRef, it must be equal too.
+func (ctrl *csiSnapshotCommonController) isVolumeBoundToClaim(volume *v1.PersistentVolume, claim *v1.PersistentVolumeClaim) bool {
```
Comment (yuxiangqian) [inline pkg/common-controller/snapshot_controller.go]: missed this call in previous rounds
I think this needs to be more restricted, i.e., if volume.Spec.ClaimRef.UID == "", it should rather . return false? the binding has not been done yet, this call is used to obtain volume information, in the case when bi-directional binding is not complete yet, it should not return true.

File: pkg/common-controller/snapshot_controller.go (hunk 20 above)
Comment (yuxiangqian) [reply pkg/common-controller/snapshot_controller.go]: Let's open an issue to resolve this, I  think the logic could be simplified, may deserve an issue to track separately.

File: pkg/common-controller/snapshot_controller.go (hunk 27 above)
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: This is strictly from the security fix.  If it is an issue, let's take a closer look and fix it separately and backport to previous releases.

**Comment by yuxiangqian** [comment]
LGTM, please create issues to track unresolved concerns.

File: pkg/common-controller/snapshot_controller.go (hunk 26 above)
Comment (xing-yang) [reply pkg/common-controller/snapshot_controller.go]: It refers to the source finalizer.  I'll just remove the "and it is not set yet" part.

**Comment by xing-yang** [comment]
> LGTM, please create issues to track unresolved concerns.

Thanks!  I'll open issues.
